its/src/test/java/com/sonar/it/scanner/msbuild/sonarcloud/IncrementalPRAnalysisSonarCloudTest.java

```diff
@@ -0,0 +1,229 @@
+
+  @ClassRule
+  public static TemporaryFolder temp = TestUtils.createTempFolder();
+
+  @Test
+  public void incrementalPrAnalysis_master() throws IOException {
+    var projectDir = TestUtils.projectDir(temp, "IncrementalPRAnalysis");
+    var logWriter = new StringWriter();
+    StreamConsumer.Pipe logsConsumer = new StreamConsumer.Pipe(logWriter);
+
+    runBeginStep(projectDir, logsConsumer);
+
+    assertThat(logWriter.toString()).contains(
+      "Processing analysis cache",
+      "Incremental PR analysis: Base branch parameter was not provided.",
+      "Cache data is empty. A full analysis will be performed.");
+  }
+
+  @Test
+  public void incrementalPrAnalysis_prWithoutChanges_producesUnchangedFilesWithAllFiles() throws IOException {
+    var projectDir = TestUtils.projectDir(temp, "IncrementalPRAnalysis");
+
+    runAnalysis(projectDir); // Initial build - master.
+    var logs = runAnalysis(projectDir, prArguments); // PR analysis.
+
+    // Verify that the file hashes are considered and all of them will be skipped.
+    assertThat(logs).contains("Incremental PR analysis: 3 files out of 3 are unchanged.");
+    Path unchangedFilesPath = getUnchangedFilesPath(projectDir);
+    assertThat(Files.readString(unchangedFilesPath)).contains("Unchanged1.cs", "Unchanged2.cs", "WithChanges.cs");
+  }
+
+  @Test
+  public void incrementalPrAnalysis_prWithChanges() throws IOException {
+    var projectDir = TestUtils.projectDir(temp, "IncrementalPRAnalysis");
+
+    runAnalysis(projectDir); // Initial build - master.
+    changeFile(projectDir, "IncrementalPRAnalysis\\WithChanges.cs"); // Change a file to force analysis.
+    runAnalysis(projectDir, prArguments); // PR analysis.
+
+    // Assert that `WithChanges.cs` file is considered modified and will be analyzed.
+    Path unchangedFilesPath = getUnchangedFilesPath(projectDir);
+    LOG.info("UnchangedFiles: " + unchangedFilesPath.toAbsolutePath());
+    assertThat(unchangedFilesPath).exists();
+    assertThat(Files.readString(unchangedFilesPath))
+      .contains("Unchanged1.cs")
+      .contains("Unchanged2.cs")
+      .doesNotContain("WithChanges.cs");
+  }
+
+  @Test
+  public void incrementalPrAnalysis_prWithChanges_basedOnDifferentBranchThanMaster() throws IOException {
+    var projectDir = TestUtils.projectDir(temp, "IncrementalPRAnalysis");
+    runAnalysis(projectDir, "/d:sonar.branch.name=different-branch"); // Initial build - different branch.
+    changeFile(projectDir, "IncrementalPRAnalysis\\WithChanges.cs"); // Change a file to force analysis.
+    // Second build - PR on top of different branch than master.
+    runAnalysis(projectDir, "/d:sonar.pullrequest.base=different-branch", "/d:sonar.pullrequest.branch=second-pull-request-branch", "/d:sonar.pullrequest.key=second-pull-request-key");
+
+    // Assert that `WithChanges.cs` file is considered modified and will be analyzed.
```

Comment:
You could extract this last part of the test as an assertion-helper method, since you re-use it in both tests that have modified files.

Reply:
Meaning this part:
```java
    // Assert that `WithChanges.cs` file is considered modified and will be analyzed.
    Path unchangedFilesPath = getUnchangedFilesPath(projectDir);
    LOG.info("UnchangedFiles: " + unchangedFilesPath.toAbsolutePath());
    assertThat(unchangedFilesPath).exists();
    assertThat(Files.readString(unchangedFilesPath))
      .contains("Unchanged1.cs")
      .contains("Unchanged2.cs")
      .doesNotContain("WithChanges.cs");
```